### PR TITLE
Quality of life tweaks

### PR DESCRIPTION
**Pull request by vddCore**
So I thought I'd add a few cents myself, since I'm using this mod like there's no tomorrow. The list of chages are as follows.

### Additional conditonal jumps
**JE/JNE** - Aliases for JZ/JNZ
**JG/JGE** - Jump if Greater / or Equal
**JL/JLE** - Jump if Lesser / or Equal

### Additional push/pop instructions
**PUSHA** - Pushes all GPRs onto stack in the following order: A B C D
**POPA** - Pops 4 values from the stack into GPRs in the following order: D C B A

### CMP instruction
The CMP instruction works like it does in x86 architecture, i.e. allows for JG/JL jumps; doesn't break existing code relying on JZ/JNZ.

New behavior of **CMP a, b**:
```
a == b -> ZF = 1, CF = 0
 a < b -> ZF = 0, CF = 1
 a > b -> ZF = 0, CF = 0
```
### Code book tweaks
Code book now has a primitive indentation support for tab key - inserts 2 spaces at the current cursor position.

### Processor fault codes
When a processor sets its fault flag it also sets a fault code that can be seen when hovering the mouse cursor over the fault flag in GUI.
image

Processor fault codes are as follows:
**0x00** - Division by zero.
**0x01** - Stack underflow (pop when stack empty).
**0x02** - Stack overflow.
**0x03** - Undefined instruction pointer location.
**0x04** - Code load error.
**0x05** - Out of bounds error.
**0xFE** - Processor stopped using HLT instruction.
**0xFF** - Processor operating correctly (default state).

### CODE REVIEW

**Comment by frodare** [comment]
These look like nice changes, thanks!  When I get a chance, I am going to look over this PR to see if I have any questions and probably merge it soon after.

File: src/main/java/net/torocraft/minecoprocessors/processor/Processor.java [hunk 1]
```diff
@@ -540,6 +569,30 @@ void processJnc() {
...
+    }
+  }
+
+  void processJl() {
+    if(carry && !zero) {
+      processJmp();
+    }
+  }
+
+  void processJle() {
```
Comment (frodare) [inline src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:622]: @Ciastex I am writing some tests cases to protect these new commands from accidental changes, and having some trouble figuring out what they do.  Can you explain them and how they are using the flags.  I am starting to think they shouldn't be added without adding a `signed` flag to the processor based on this: http://faydoc.tripod.com/cpu/jg.htm

File: src/main/java/net/torocraft/minecoprocessors/processor/Processor.java [hunk 2]
```diff
@@ -540,6 +601,30 @@ void processJnc() {
...
   }
 
+  void processJg() {
+    if(!carry && !zero) {
+      processJmp();
+    }
+  }
+
+  void processJge() {
+    if((!carry && !zero) || (!carry && zero)) {
```
Comment (zeriley) [inline src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:611]: Couldn't this just be:
if(!carry) {
?

File: src/main/java/net/torocraft/minecoprocessors/processor/Processor.java [hunk 3]
```diff
@@ -540,6 +601,30 @@ void processJnc() {
...
+  }
+
+  void processJl() {
+    if(carry && !zero) {
+      processJmp();
+    }
+  }
+
+  void processJle() {
+    if((carry && zero) || (!carry && zero)) {
```
Comment (zeriley) [inline src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:623]: And this one could just be:
if (zero) {
?

File: src/main/java/net/torocraft/minecoprocessors/processor/Processor.java (hunk 1 above)
Comment (vddCore) [reply src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:622]: I misnamed the instructions. x86 JA/JB/JAE/JBE work based on ZF/CF and perform an unsigned comparison. I tried to figure out the simplest way of implementing the comparison of two numbers using flags that are already provided. I used the description below and based the jumps according to the CMP instruction modification: https://www.aldeid.com/wiki/X86-assembly/Instructions/cmp

Comment (frodare) [reply src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:622]: I am going keep them as is for now, and update them and many other instructions if or when a SF and OF (maybe even a PF) are added

File: src/main/java/net/torocraft/minecoprocessors/processor/Processor.java (hunk 2 above)
Comment (frodare) [reply src/main/java/net/torocraft/minecoprocessors/processor/Processor.java:611]: I was thinking the same thing, but it was hurting my head last night :face_with_head_bandage:

**Comment by frodare** [comment]
@Ciastex I have uploaded these changes to curse, version 1.12.2-5, thanks again!

**Pull request merged**